super_editor/lib/src/default_editor/document_caret_overlay.dart

```diff
@@ -0,0 +1,134 @@
+import 'package:flutter/widgets.dart';
+import 'package:super_editor/src/core/document_composer.dart';
+import 'package:super_editor/src/core/document_layout.dart';
+import 'package:super_text_layout/super_text_layout.dart';
+
+/// Document overlay that paints a caret with the given [caretStyle].
+class CaretDocumentOverlay extends StatefulWidget {
+  const CaretDocumentOverlay({
+    Key? key,
+    required this.composer,
+    required this.documentLayoutResolver,
+    required this.caretStyle,
+  }) : super(key: key);
+
+  /// The editor's [DocumentComposer], which reports the current selection.
+  final DocumentComposer composer;
+
+  /// Delegate that returns a reference to the editor's [DocumentLayout], so
+  /// that the current selection can be mapped to an (x,y) offset and a height.
+  final DocumentLayout Function() documentLayoutResolver;
+
+  /// The visual style of the caret that this overlay paints.
+  final CaretStyle caretStyle;
+
+  @override
+  State<CaretDocumentOverlay> createState() => _CaretDocumentOverlayState();
+}
+
+class _CaretDocumentOverlayState extends State<CaretDocumentOverlay> with SingleTickerProviderStateMixin {
+  final _caret = ValueNotifier<Rect?>(null);
+  late final BlinkController _blinkController;
+  BoxConstraints? _previousConstraints;
+
+  @override
+  void initState() {
+    super.initState();
+    widget.composer.selectionNotifier.addListener(_onSelectionChange);
+    _blinkController = BlinkController(tickerProvider: this)..startBlinking();
+  }
+
+  @override
+  void didUpdateWidget(CaretDocumentOverlay oldWidget) {
+    super.didUpdateWidget(oldWidget);
+
+    if (widget.composer != oldWidget.composer) {
+      oldWidget.composer.removeListener(_onSelectionChange);
+      widget.composer.addListener(_onSelectionChange);
```

Comment:
Shouldn't we remove/re-add a listener to `selectionNotifier`?
```suggestion
      oldWidget.composer.selectionNotifier.removeListener(_onSelectionChange);
      widget.composer.selectionNotifier.addListener(_onSelectionChange);
```